src/xrpld/app/tx/detail/InvariantCheck.cpp

```diff
@@ -2170,4 +2179,911 @@ ValidAMM::finalize(
+                                   "vault must have no assets outstanding";
+                result = false;
+            }
+            if (afterVault_->assetsAvailable != zero)
+            {
+                JLOG(j.fatal()) << "Invariant failed: updated zero sized "
+                                   "vault must have no assets available";
+                result = false;
+            }
+        }
+        else if (updatedShares->sharesTotal > updatedShares->sharesMaximum)
+        {
+            JLOG(j.fatal())  //
+                << "Invariant failed: updated shares must not exceed maximum "
+                << updatedShares->sharesMaximum;
+            result = false;
+        }
+
+        if (afterVault_->assetsAvailable < zero)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: assets available must be positive";
+            result = false;
+        }
+
+        if (afterVault_->assetsAvailable > afterVault_->assetsTotal)
+        {
+            JLOG(j.fatal()) << "Invariant failed: assets available must "
+                               "not be greater than assets outstanding";
+            result = false;
+        }
+        else if (
+            afterVault_->lossUnrealized >
+            afterVault_->assetsTotal - afterVault_->assetsAvailable)
+        {
+            JLOG(j.fatal())  //
+                << "Invariant failed: loss unrealized must not exceed "
+                   "the difference between assets outstanding and available";
+            result = false;
+        }
+
+        if (afterVault_->assetsTotal < zero)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: assets outstanding must be positive";
+            result = false;
+        }
+
+        if (afterVault_->assetsMaximum < zero)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: assets maximum must be positive";
+            result = false;
+        }
+    }
+
+    if (txnType == ttVAULT_CREATE ||    //
+        txnType == ttVAULT_SET ||       //
+        txnType == ttVAULT_DEPOSIT ||   //
+        txnType == ttVAULT_WITHDRAW ||  //
+        txnType == ttVAULT_CLAWBACK)
+    {
+        if (!afterVault_)
+        {
+            JLOG(j.fatal()) <<  //
+                "Invariant failed: vault operation succeeded without updating "
+                "or creating a vault";
+            XRPL_ASSERT(
+                enforce, "ripple::ValidVault::finalize : vault noop invariant");
+            return !enforce;  // That's all we can do here
+        }
+
+        // Transactor makes this condition impossible, but since we need to
+        // access beforeVault_ then we also need a defensive check.
+        if (!beforeVault_ && txnType != ttVAULT_CREATE)
+        {
+            // LCOV_EXCL_START
+            UNREACHABLE(
+                "ripple::ValidVault::finalize : missing old vault state");
+            return false;
+            // LCOV_EXCL_STOP
+        }
```

Comment:
Remember, the whole purpose of the invariants are to catch things that might be wrong in transactors, so something like "Transactor makes this condition impossible" is the wrong way to look at this. _All_ of these failure conditions should be impossible.

Also, this is absolutely testable using the framework in `Invariants_test.cpp`. Do a test case that manually creates, modifies, or deletes a vault, but have the transaction type be, I dunno, `ttPayment`.